File: crates/control-plane/src/handlers/events.rs
//! Event handling API handlers.
//!
//! Handles worker events and command retrieval endpoints.
//!
//! SECURITY: All event payloads are sanitized before storage to prevent
//! sensitive data (bearer tokens, passwords, API keys) from being persisted.

use axum::{
    extract::{Path, State},
    Json,
};
use serde::{Deserialize, Serialize};
use tracing::{debug, info, warn};

use crate::error::{AppError, AppResult};
use crate::sanitize::sanitize_sensitive_data;
use crate::state::AppState;

/// Worker event request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EventRequest {
    /// Execution ID.
    pub execution_id: String,
    /// Step name.
    pub step: String,
    /// Event name (step.enter, call.done, step.exit, etc.).
    pub name: String,
    /// Event payload/result data.
    #[serde(default)]
    pub payload: serde_json::Value,
    /// Additional metadata.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meta: Option<serde_json::Value>,
    /// Worker ID.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub worker_id: Option<String>,
    /// Result kind: "data", "ref", or "refs".
    #[serde(default = "default_result_kind")]
    pub result_kind: String,
    /// Result URI for ref kind.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result_uri: Option<String>,
    /// Event IDs for refs kind.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub event_ids: Option<Vec<i64>>,
    /// If true, server should take action.
    #[serde(default = "default_true")]
    pub actionable: bool,
    /// If true, event is for logging/observability.
    #[serde(default = "default_true")]
    pub informative: bool,
}

fn default_result_kind() -> String {
    "data".to_string()
}

fn default_true() -> bool {
    true
}

/// Response for event handling.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EventResponse {
    /// Status of the operation.
    pub status: String,
    /// Event ID that was created.
    pub event_id: i64,
    /// Number of commands generated.
    pub commands_generated: i32,
}

/// Command details response.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandResponse {
    /// Execution ID.
    pub execution_id: i64,
    /// Node/step ID.
    pub node_id: String,
    /// Node/step name.
    pub node_name: String,
    /// Action/tool kind.
    pub action: String,
    /// Command context (tool config, args, etc.).
    pub context: serde_json::Value,
    /// Command metadata.
    pub meta: serde_json::Value,
}

/// Handle worker event.
///
/// POST /api/events
///
/// Worker reports completion with result (inline or ref).
/// Engine evaluates case/when/then and generates next commands.
pub async fn handle_event(
    State(state): State<AppState>,
    Json(request): Json<EventRequest>,
) -> Result<Json<EventResponse>, AppError> {
    debug!(
        "Event received: execution_id={}, step={}, name={}",
        request.execution_id, request.step, request.name
    );

    let execution_id: i64 = request
        .execution_id
        .parse()
        .map_err(|_| AppError::Validation("Invalid execution_id".to_string()))?;

    // Events that should NOT trigger engine processing
    let skip_engine_events = [
        "command.claimed",
        "command.started",
        "command.completed",
        "command.failed",
        "step.enter",
    ];

    // For command.claimed, check if already claimed
    if request.name == "command.claimed" {
        if let Some(command_id) = get_command_id(&request) {
            if check_already_claimed(&state, execution_id, &command_id, &request.worker_id).await? {
                // Already claimed by same worker - idempotent success
                return Ok(Json(EventResponse {
                    status: "ok".to_string(),
                    event_id: 0,
                    commands_generated: 0,
                }));
            }
        }
    }

    // Build result object based on kind
    let result_obj_raw = build_result_object(&request);
    // SECURITY: Sanitize result data to remove sensitive information (tokens, passwords, etc.)
    let result_obj = sanitize_sensitive_data(&result_obj_raw);

    // Generate event ID
    let event_id = generate_snowflake_id(&state).await?;

    // Get catalog_id from existing events
    let catalog_id = get_catalog_id(&state, execution_id).await?;

    // Build meta object with control flags
    let mut meta_obj = request.meta.clone().unwrap_or(serde_json::json!({}));
    if let serde_json::Value::Object(ref mut map) = meta_obj {
        map.insert(
            "actionable".to_string(),
            serde_json::json!(request.actionable),
        );
        map.insert(
            "informative".to_string(),
            serde_json::json!(request.informative),
        );
        if let Some(ref worker_id) = request.worker_id {
            map.insert("worker_id".to_string(), serde_json::json!(worker_id));
        }
    }
    // SECURITY: Sanitize meta data to remove sensitive information
    let meta_obj = sanitize_sensitive_data(&meta_obj);

    // Determine status based on event name
    let status = if request.name.contains("done") || request.name.contains("exit") {
        "COMPLETED"
    } else if request.name.contains("error") || request.name.contains("failed") {
        "FAILED"
    } else {
        "RUNNING"
    };

    // Persist the event
    sqlx::query(
        r#"
        INSERT INTO noetl.event (
            event_id, execution_id, catalog_id, event_type,
            node_id, node_name, status, result, meta, created_at
        ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
        "#,
    )
    .bind(event_id)
    .bind(execution_id)
    .bind(catalog_id)
    .bind(&request.name)
    .bind(&request.step)
    .bind(&request.step)
    .bind(status)
    .bind(&result_obj)
    .bind(&meta_obj)
    .bind(chrono::Utc::now())
    .execute(&state.db)
    .await?;

    info!(
        "Event persisted: event_id={}, execution_id={}, name={}",
        event_id, execution_id, request.name
    );

    // Process through engine if applicable
    let commands_generated = if !skip_engine_events.contains(&request.name.as_str()) {
        // TODO: Implement engine event handling
        // This would call the orchestrator to evaluate next steps
        debug!("Would process through engine: event_type={}", request.name);
        0
    } else {
        debug!("Skipped engine for administrative event: {}", request.name);
        0
    };

    // Trigger orchestrator for workflow progression on command.completed
    if request.name == "command.completed" && request.step.to_lowercase() != "end" {
        match trigger_orchestrator(&state, execution_id, event_id).await {
            Ok(cmds) => {
                info!(
                    "Orchestrator generated {} commands for execution {}",
                    cmds, execution_id
                );
            }
            Err(e) => {
                warn!("Orchestrator error: {}", e);
            }
        }
    }

    Ok(Json(EventResponse {
        status: "ok".to_string(),
        event_id,
        commands_generated,
    }))
}

/// Get command details from command.issued event.
///
/// GET /api/commands/{event_id}
///
/// Workers call this to fetch command config after NATS notification.
pub async fn get_command(
    State(state): State<AppState>,
    Path(event_id): Path<i64>,
) -> Result<Json<CommandResponse>, AppError> {
    debug!("Getting command for event_id={}", event_id);

    let row: Option<(i64, String, String, serde_json::Value, serde_json::Value)> =
        sqlx::query_as::<_, (i64, String, String, serde_json::Value, serde_json::Value)>(
            r#"
            SELECT execution_id, node_name, node_type, context, meta
            FROM noetl.event
            WHERE event_id = $1 AND event_type = 'command.issued'
            "#,
        )
        .bind(event_id)
        .fetch_optional(&state.db)
        .await?;

    match row {
        Some((execution_id, node_name, node_type, context, meta)) => Ok(Json(CommandResponse {
            execution_id,
            node_id: node_name.clone(),
            node_name,
            action: node_type,
            context,
            meta,
        })),
        None => Err(AppError::NotFound(format!(
            "command.issued event not found: {}",
            event_id
        ))),
    }
}

/// Extract command_id from request.
fn get_command_id(request: &EventRequest) -> Option<String> {
    // Try payload first
    if let Some(id) = request.payload.get("command_id").and_then(|v| v.as_str()) {
        return Some(id.to_string());
    }
    // Try meta
    if let Some(meta) = &request.meta {
        if let Some(id) = meta.get("command_id").and_then(|v| v.as_str()) {
            return Some(id.to_string());
        }
    }
    None
}

/// Check if command is already claimed.
async fn check_already_claimed(
    state: &AppState,
    execution_id: i64,
    command_id: &str,
    worker_id: &Option<String>,
) -> AppResult<bool> {
    let row: Option<(Option<String>, Option<serde_json::Value>)> =
        sqlx::query_as::<_, (Option<String>, Option<serde_json::Value>)>(
            r#"
            SELECT worker_id, meta FROM noetl.event
            WHERE execution_id = $1
              AND event_type = 'command.claimed'
              AND (meta->>'command_id' = $2 OR result->'data'->>'command_id' = $2)
            LIMIT 1
            "#,
        )
        .bind(execution_id)
        .bind(command_id)
        .fetch_optional(&state.db)
        .await?;

    if let Some((existing_worker, meta)) = row {
        let existing_worker_id = existing_worker.or_else(|| {
            meta.and_then(|m| {
                m.get("worker_id")
                    .and_then(|v| v.as_str())
                    .map(String::from)
            })
        });

        if let (Some(existing), Some(current)) = (&existing_worker_id, worker_id) {
            if existing != current {
                // Different worker - reject
                return Err(AppError::Conflict(format!(
                    "Command already claimed by {}",
                    existing
                )));
            }
            // Same worker - idempotent
            return Ok(true);
        }
    }

    Ok(false)
}

/// Build result object based on result_kind.
fn build_result_object(request: &EventRequest) -> serde_json::Value {
    match request.result_kind.as_str() {
        "ref" if request.result_uri.is_some() => {
            let uri = request.result_uri.as_ref().unwrap();
            let store_tier = if uri.starts_with("gs://") {
                "gcs"
            } else if uri.starts_with("s3://") {
                "s3"
            } else {
                "artifact"
            };
            serde_json::json!({
                "kind": "ref",
                "store_tier": store_tier,
                "logical_uri": uri,
            })
        }
        "refs" if request.event_ids.is_some() => {
            let event_ids = request.event_ids.as_ref().unwrap();
            serde_json::json!({
                "kind": "refs",
                "event_ids": event_ids,
                "total_parts": event_ids.len(),
            })
        }
        _ => {
            serde_json::json!({
                "kind": "data",
                "data": request.payload,
            })
        }
    }
}

/// Get catalog_id from existing events.
async fn get_catalog_id(state: &AppState, execution_id: i64) -> AppResult<Option<i64>> {
    let row: Option<(i64,)> = sqlx::query_as::<_, (i64,)>(
        "SELECT catalog_id FROM noetl.event WHERE execution_id = $1 LIMIT 1",
    )
    .bind(execution_id)
    .fetch_optional(&state.db)
    .await?;

    Ok(row.map(|(id,)| id))
}

/// Generate a snowflake ID.
async fn generate_snowflake_id(state: &AppState) -> AppResult<i64> {
    let row: (i64,) = sqlx::query_as::<_, (i64,)>("SELECT noetl.snowflake_id()")
        .fetch_one(&state.db)
        .await?;

    Ok(row.0)
}

/// Trigger orchestrator for workflow progression.
async fn trigger_orchestrator(
    _state: &AppState,
    execution_id: i64,
    trigger_event_id: i64,
) -> AppResult<i32> {
    // This would be a full orchestrator implementation
    // For now, just log and return 0
    debug!(
        "Would trigger orchestrator for execution={}, trigger_event={}",
        execution_id, trigger_event_id
    );

    // TODO: Load playbook, reconstruct state from events, evaluate next steps
    Ok(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_event_request_defaults() {
        let json = r#"{"execution_id": "123", "step": "step1", "name": "step.enter"}"#;
        let request: EventRequest = serde_json::from_str(json).unwrap();

        assert_eq!(request.result_kind, "data");
        assert!(request.actionable);
        assert!(request.informative);
    }

    #[test]
    fn test_build_result_object_data() {
        let request = EventRequest {
            execution_id: "123".to_string(),
            step: "step1".to_string(),
            name: "step.exit".to_string(),
            payload: serde_json::json!({"output": "success"}),
            meta: None,
            worker_id: None,
            result_kind: "data".to_string(),
            result_uri: None,
            event_ids: None,
            actionable: true,
            informative: true,
        };

        let result = build_result_object(&request);
        assert_eq!(result["kind"], "data");
        assert_eq!(result["data"]["output"], "success");
    }

    #[test]
    fn test_build_result_object_ref() {
        let request = EventRequest {
            execution_id: "123".to_string(),
            step: "step1".to_string(),
            name: "step.exit".to_string(),
            payload: serde_json::json!({}),
            meta: None,
            worker_id: None,
            result_kind: "ref".to_string(),
            result_uri: Some("gs://bucket/path/to/result".to_string()),
            event_ids: None,
            actionable: true,
            informative: true,
        };

        let result = build_result_object(&request);
        assert_eq!(result["kind"], "ref");
        assert_eq!(result["store_tier"], "gcs");
        assert_eq!(result["logical_uri"], "gs://bucket/path/to/result");
    }

    #[test]
    fn test_event_response_serialization() {
        let response = EventResponse {
            status: "ok".to_string(),
            event_id: 12345,
            commands_generated: 2,
        };

        let json = serde_json::to_string(&response).unwrap();
        assert!(json.contains("ok"));
        assert!(json.contains("12345"));
    }

    #[test]
    fn test_command_response_serialization() {
        let response = CommandResponse {
            execution_id: 12345,
            node_id: "step1".to_string(),
            node_name: "step1".to_string(),
            action: "python".to_string(),
            context: serde_json::json!({"tool_config": {}}),
            meta: serde_json::json!({"attempt": 1}),
        };

        let json = serde_json::to_string(&response).unwrap();
        assert!(json.contains("step1"));
        assert!(json.contains("python"));
    }
}

File: noetl/server/api/v2/events.py
import os
import json
import time
import asyncio
import heapq
import math
from dataclasses import dataclass
from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field, model_validator
from typing import Any, Optional, Literal
from datetime import datetime, timezone
from psycopg.types.json import Json
from psycopg.rows import dict_row
from psycopg_pool import PoolTimeout
from noetl.core.dsl.v2.models import Event
from noetl.core.dsl.v2.engine import ControlFlowEngine, PlaybookRepo, StateStore
from noetl.core.db.pool import get_pool_connection, get_server_pool_stats
from noetl.core.messaging import NATSCommandPublisher
from noetl.core.storage import Scope, default_store, estimate_size
from noetl.claim_policy import decide_reclaim_for_existing_claim
from noetl.server.api.event_queries import PENDING_COMMAND_COUNT_SQL
from noetl.server.api.supervision import supervise_command_issued, supervise_persisted_event
from noetl.core.logger import setup_logger

from .core import *
from .models import *
from .utils import *
from .cache import *
from .db import *
from .recovery import *
from .commands import *
@router.post('/events', response_model=EventResponse)
async def handle_event(req: EventRequest) -> EventResponse:
    """
    Handle worker event.
    
    Worker reports completion with result (inline or ref).
    Engine evaluates case/when/then and generates next commands.
    
    CRITICAL: Only process through engine for events that drive workflow:
    - step.exit: Step completed, evaluate case rules and generate next commands
    - call.done: Action completed, may trigger case rules
    - call.error: Action failed, may trigger error handling
    - command.failed: Command-level failure, must emit terminal failure lifecycle if unhandled
    - loop.item/loop.done: Loop iteration events
    
    Skip engine for administrative events:
    - command.claimed: Just persist, don't process
    - command.heartbeat: lease/liveness refresh, don't process
    - command.started: Just persist, don't process
    - command.completed: Already processed by worker
    - step.enter: Just marks step started
    """
    engine: Optional[ControlFlowEngine] = None
    commands_generated = False
    try:
        _raise_if_db_short_circuit_enabled(operation='handle_event')
        engine = get_engine()
        skip_engine_events = {'command.claimed', 'command.heartbeat', 'command.started', 'command.completed', 'step.enter'}
        if req.name == 'command.claimed':
            try:
                _validate_reference_only_payload(req.payload)
            except ValueError as exc:
                raise HTTPException(status_code=422, detail=f"Invalid reference-only event payload for step '{req.step}' event '{req.name}': {exc}") from exc
            command_id = req.payload.get('command_id') or (req.meta or {}).get('command_id')
            if command_id:
                async with get_pool_connection() as conn:
                    async with conn.cursor(row_factory=dict_row) as cur:
                        evt_id = await _next_snowflake_id(cur)
                        await cur.execute('\n                            SELECT pg_try_advisory_xact_lock(hashtext(%s)::bigint) as lock_acquired\n                        ', (command_id,))
                        lock_result = await cur.fetchone()
                        if not lock_result or not lock_result.get('lock_acquired'):
                            logger.warning(f'[CLAIM-REJECT] Command {command_id} lock held by another worker, rejecting claim from {req.worker_id}')
                            raise HTTPException(409, f'Command already being claimed by another worker')
                        await cur.execute(_HANDLE_EVENT_CLAIMED_LOOKUP_SQL, _command_id_lookup_params(int(req.execution_id), command_id))
                        existing = await cur.fetchone()
                        if existing:
                            existing_worker = existing.get('worker_id') or (existing.get('meta') or {}).get('worker_id')
                            if existing_worker and existing_worker != req.worker_id:
                                logger.warning(f'[CLAIM-REJECT] Command {command_id} already claimed by {existing_worker}, rejecting claim from {req.worker_id}')
                                raise HTTPException(409, f'Command already claimed by {existing_worker}')
                            else:
                                logger.info(f'[CLAIM-IDEMPOTENT] Command {command_id} already claimed by SAME worker {req.worker_id}, returning success')
                                return EventResponse(status='ok', event_id=0, commands_generated=0)
                        await cur.execute('SELECT catalog_id FROM noetl.event WHERE execution_id = %s LIMIT 1', (int(req.execution_id),))
                        row = await cur.fetchone()
                        catalog_id = row['catalog_id'] if row else None
                        meta_obj = dict(req.meta or {})
                        meta_obj['actionable'] = req.actionable
                        meta_obj['informative'] = req.informative
                        meta_obj['command_id'] = command_id
                        if req.worker_id:
                            meta_obj['worker_id'] = req.worker_id
                        result_obj = _build_reference_only_result(payload=req.payload, status='RUNNING')
                        await cur.execute('\n                            INSERT INTO noetl.event (\n                                event_id, execution_id, catalog_id, event_type,\n                                node_id, node_name, status, result, meta, worker_id, created_at\n                            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\n                        ', (evt_id, int(req.execution_id), catalog_id, req.name, req.step, req.step, 'RUNNING', Json(result_obj), Json(meta_obj), req.worker_id, datetime.now(timezone.utc)))
                        await conn.commit()
                        _record_db_operation_success()
                        logger.info(f'[CLAIM-SUCCESS] Command {command_id} claimed by worker {req.worker_id}')
                        return EventResponse(status='ok', event_id=evt_id, commands_generated=0)
        status = _status_from_event_name(req.name)
        try:
            _validate_reference_only_payload(req.payload)
        except ValueError as exc:
            raise HTTPException(status_code=422, detail=f"Invalid reference-only event payload for step '{req.step}' event '{req.name}': {exc}") from exc
        result_obj = _build_reference_only_result(payload=req.payload, status=status)
        error_text = _extract_event_error(req.payload)
        command_id = _extract_event_command_id(req)
        event_meta = dict(req.meta or {})
        event_meta['actionable'] = req.actionable
        event_meta['informative'] = req.informative
        if req.worker_id:
            event_meta['worker_id'] = req.worker_id
        if command_id:
            event_meta['command_id'] = command_id
        async with get_pool_connection() as conn:
            async with conn.cursor(row_factory=dict_row) as cur:
                await cur.execute('SELECT catalog_id FROM noetl.event WHERE execution_id = %s LIMIT 1', (int(req.execution_id),))
                row = await cur.fetchone()
                catalog_id = row['catalog_id'] if row else None
                if command_id and req.name in _COMMAND_EVENT_DEDUPE_TYPES:
                    await cur.execute("\n                        SELECT event_id\n                        FROM noetl.event\n                        WHERE execution_id = %s\n                          AND event_type = %s\n                          AND node_name = %s\n                          AND meta ? 'command_id'\n                          AND meta->>'command_id' = %s\n                        ORDER BY event_id DESC\n                        LIMIT 1\n                        ", (int(req.execution_id), req.name, req.step, command_id))
                    duplicate_row = await cur.fetchone()
                    if duplicate_row:
                        duplicate_event_id = int(duplicate_row.get('event_id'))
                        logger.warning('[EVENT-DEDUPE] Ignoring duplicate event %s for execution=%s step=%s command_id=%s existing_event_id=%s', req.name, req.execution_id, req.step, command_id, duplicate_event_id)
                        if req.name in {'command.completed', 'command.failed'}:
                            _active_claim_cache_invalidate(command_id=command_id)
                        return EventResponse(status='ok', event_id=duplicate_event_id, commands_generated=0)
                evt_id = await _next_snowflake_id(cur)
                event_meta['persisted_event_id'] = str(evt_id)
                await cur.execute('\n                    INSERT INTO noetl.event (\n                        event_id, execution_id, catalog_id, event_type,\n                        node_id, node_name, status, result, meta, error, created_at\n                    ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\n                ', (evt_id, int(req.execution_id), catalog_id, req.name, req.step, req.step, status, Json(result_obj), Json(event_meta), error_text, datetime.now(timezone.utc)))
                await conn.commit()
                _record_db_operation_success()
        await supervise_persisted_event(req.execution_id, req.step, req.name, req.payload, event_meta, event_id=int(evt_id))
        if req.name in {'command.completed', 'command.failed'}:
            if command_id:
                _active_claim_cache_invalidate(command_id=command_id)
        event = Event(execution_id=req.execution_id, step=req.step, name=req.name, payload=req.payload, meta=event_meta, timestamp=datetime.now(timezone.utc), worker_id=req.worker_id, attempt=event_meta.get('attempt', 1))
        commands = []
        if req.name not in skip_engine_events:
            async with get_pool_connection() as engine_conn:
                async with engine_conn.transaction():
                    async with engine_conn.cursor() as cur:
                        await cur.execute('SELECT pg_advisory_xact_lock(%s)', (int(req.execution_id),))
                    commands = await engine.handle_event(event, conn=engine_conn, already_persisted=True)
            commands_generated = bool(commands)
            logger.debug(f'[ENGINE] Processed {req.name} for step {req.step}, generated {len(commands)} commands')
        else:
            logger.debug(f'[ENGINE] Skipped engine for administrative event {req.name}')
        server_url = os.getenv('NOETL_SERVER_URL', 'http://noetl.noetl.svc.cluster.local:8082')
        command_events = []
        supervisor_commands: list[tuple[str, str, str, int, dict[str, Any]]] = []
        async with get_pool_connection() as conn:
            async with conn.cursor(row_factory=dict_row) as cur:
                for cmd in commands:
                    await cur.execute('SELECT catalog_id, parent_execution_id FROM noetl.event WHERE execution_id = %s LIMIT 1', (int(cmd.execution_id),))
                    row = await cur.fetchone()
                    cat_id = row['catalog_id'] if row else catalog_id
                    parent_exec = row['parent_execution_id'] if row else None
                    cmd_suffix = await _next_snowflake_id(cur)
                    cmd_id = f'{cmd.execution_id}:{cmd.step}:{cmd_suffix}'
                    new_evt_id = await _next_snowflake_id(cur)
                    context = _build_command_context(cmd)
                    _validate_postgres_command_context_or_422(step=cmd.step, tool_kind=cmd.tool.kind, context=context)
                    meta = {'command_id': cmd_id, 'step': cmd.step, 'tool_kind': cmd.tool.kind, 'triggered_by': req.name, 'trigger_step': req.step, 'actionable': True}
                    if cmd.metadata:
                        meta.update({k: v for k, v in cmd.metadata.items() if v is not None})
                    context = await _store_command_context_if_needed(execution_id=int(cmd.execution_id), step=cmd.step, command_id=cmd_id, context=context)
                    await cur.execute('\n                        INSERT INTO noetl.event (\n                            event_id, execution_id, catalog_id, event_type,\n                            node_id, node_name, node_type, status,\n                            context, meta, parent_event_id, parent_execution_id,\n                            created_at\n                        ) VALUES (\n                            %(event_id)s, %(execution_id)s, %(catalog_id)s, %(event_type)s,\n                            %(node_id)s, %(node_name)s, %(node_type)s, %(status)s,\n                            %(context)s, %(meta)s, %(parent_event_id)s, %(parent_execution_id)s,\n                            %(created_at)s\n                        )\n                    ', {'event_id': new_evt_id, 'execution_id': int(cmd.execution_id), 'catalog_id': cat_id, 'event_type': 'command.issued', 'node_id': cmd.step, 'node_name': cmd.step, 'node_type': cmd.tool.kind, 'status': 'PENDING', 'context': Json(context), 'meta': Json(meta), 'parent_event_id': evt_id, 'parent_execution_id': parent_exec, 'created_at': datetime.now(timezone.utc)})
                    command_events.append((int(cmd.execution_id), new_evt_id, cmd_id, cmd.step))
                    supervisor_commands.append((str(cmd.execution_id), cmd_id, cmd.step, int(new_evt_id), dict(meta)))
                await conn.commit()
        for supervisor_execution_id, supervisor_command_id, supervisor_step, supervisor_event_id, supervisor_meta in supervisor_commands:
            await supervise_command_issued(supervisor_execution_id, supervisor_command_id, supervisor_step, event_id=supervisor_event_id, meta=supervisor_meta)
        await _publish_commands_with_recovery(command_events, server_url=server_url)
        if req.name == 'command.completed' and req.step.lower() != 'end':
            try:
                from .run.orchestrator import evaluate_execution
                await evaluate_execution(execution_id=str(req.execution_id), trigger_event_type='command.completed', trigger_event_id=str(evt_id))
            except ImportError:
                logger.debug('Orchestrator module not available, skipping for command.completed')
            except Exception as e:
                logger.warning(f'Orchestrator error: {e}')
        terminal_events = {'playbook.completed', 'playbook.failed', 'workflow.completed', 'workflow.failed', 'execution.cancelled'}
        if req.name in terminal_events:
            try:
                await engine.state_store.evict_completed(req.execution_id)
                logger.debug(f'Evicted execution {req.execution_id} from cache after {req.name}')
            except Exception as e:
                logger.warning(f'Failed to evict execution {req.execution_id} from cache: {e}')
        return EventResponse(status='ok', event_id=evt_id, commands_generated=len(commands))
    except PoolTimeout:
        if engine is not None and commands_generated:
            await _invalidate_execution_state_cache(req.execution_id, reason='command_issue_pool_timeout', engine=engine)
        retry_after = _compute_retry_after()
        logger.warning('[EVENTS] DB pool saturated while persisting %s for step %s retry_after=%ss', req.name, req.step, retry_after)
        raise HTTPException(status_code=503, detail={'code': 'pool_saturated', 'message': 'Database temporarily overloaded; retry shortly'}, headers={'Retry-After': retry_after})
    except Exception as e:
        if engine is not None and commands_generated:
            await _invalidate_execution_state_cache(req.execution_id, reason=f'command_issue_failed:{type(e).__name__}', engine=engine)
        retry_after = _record_db_unavailable_failure(e, operation='handle_event')
        if retry_after is not None:
            raise HTTPException(status_code=503, detail={'code': 'db_unavailable', 'message': 'Database temporarily unavailable; retry shortly'}, headers={'Retry-After': retry_after})
        logger.error(f'handle_event failed: {e}', exc_info=True)
        raise HTTPException(500, str(e))

__all__ = ['handle_event']
